flow based tablet load balancer

demmer opened the pull request:
<!--
  Thank you for your contribution to the Vitess project.
  How to contribute: https://vitess.io/docs/contributing/
  Please first make sure there is an open Issue to discuss the feature/fix suggested in this PR.
  If this is a new feature, please mark the Issue as "RFC".
 -->

<!-- if this PR is Work in Progress please create it as a Draft Pull Request -->

## Description
Implement an optional flow-based tablet balancer. 

The motivation and algorithm is described more in https://github.com/vitessio/vitess/issues/12241 and in a long comment in the code.

In brief, the goal of this implementation is to more evenly balance out the queries to a the replicas in a shard when the tablets are not evenly distributed across the various cells where vtgates are receiving application traffic, because the default routing algorithm will always prefer an available tablet in the same cell as the vtgate, even if there are more idle tablets in other cells. The balancer addresses this by attempting to route an equal share to each tablet, preferring the same-cell where possible, but crossing cells in order to achieve better global balance.

## Slack production results
Note that the original algorithm was developed many months ago, but only recently have we made the time to deploy this within Slack and spent some time to verify it works as expected.

The results from our initial testing are very encouraging and we have been running with this feature enabled for some time now on one of our main production keyspaces. This is a 16 shard keyspace with between 9-11 tablets per shard spread among our 4 active cells. Each cell receives the same amount of inbound traffic.

When we first enabled the feature shows that the wide disparity in QPS per replica goes away and instead the QPS is only based on the number of replicas per shard, as expected:
![image](https://github.com/vitessio/vitess/assets/3308504/39899f9b-9497-42fa-a9ab-dc9ffd5279ff)

Since then we have actually normalized our deployment so each shard has 9 replicas and the QPS is quite even across all hosts, even though again they do not map evenly to each cell:
<img width="1629" alt="image" src="https://github.com/vitessio/vitess/assets/3308504/ea0d7662-dc24-40c7-944a-bf6dbce6789f">

## Related Issue(s)
https://github.com/vitessio/vitess/issues/12241

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required https://github.com/vitessio/website/pull/1804

## Deployment Notes

<!-- Notes regarding deployment of the contained body of work. These should note any db migrations, etc. -->

systay commented:
This looks great @demmer 

Just need to use `rand/v2` to make CI happy, and then we need a website PR for the new flags, right?

demmer commented:
@systay based on the test coverage reports, @venkatraju and I added some more tests for the balancer, and I beefed up some of the existing tests for the retry logic, including running both with and without the balancer. might be overkill but this is kinda a critical part of the code :)

deepthi commented on go/vt/vtgate/tabletgateway.go:
```diff
@@ -62,6 +70,9 @@ func init() {
 		fs.DurationVar(&initialTabletTimeout, "gateway_initial_tablet_timeout", 30*time.Second, "At startup, the tabletGateway will wait up to this duration to get at least one tablet per keyspace/shard/tablet type")
 		fs.IntVar(&retryCount, "retry-count", 2, "retry count")
+		fs.BoolVar(&balancerEnabled, "balancer_enabled", false, "Whether to enable the tablet balancer to evenly spread query load")
+		fs.StringSliceVar(&balancerVtgateCells, "balancer_vtgate_cells", []string{}, "When in balanced mode, a comma-separated list of cells that contain vtgates (required)")
```
Everything else is cosmetic, but this seems important.
What happens if you add vtgates to a cell that didn't previously have any? You would have to restart every vtgate with a new value for this flag?
It may not be very common in practice, but it should be documented somewhere.
Of course, the real answer is that vtgates should be given a list of all cells and discover which ones contain vtgates, but we've so far avoided adding vtgates to the topo.

demmer replied:
>What happens if you add vtgates to a cell that didn't previously have any? You would have to restart every vtgate with a new value for this flag?

Yep. For the current algorithm to work as designed, it's required that all the vtgates know the fully set of cells that are running vtgates and that the load among these cells is roughly even. This allows each individual vtgate to make the assumption about what the others will do without needing signaling.

If you want to change this assumption therefore, you do have to restart the whole fleet to take it into account, just like any other vtgate flag change.

> It may not be very common in practice, but it should be documented somewhere.

That makes sense. I'll include something to this effect when we write the website docs.

demmer replied:
> Of course, the real answer is that vtgates should be given a list of all cells and discover which ones contain vtgates, but we've so far avoided adding vtgates to the topo.

Yeah - that's obviously beyond the scope of this PR, and it would in theory reduce the dependency that query inflows per cell need to be balanced (i.e. you could use the number of gates as a proxy for how much load you expect).

demmer commented on go/vt/vtgate/tabletgateway.go:
```diff
@@ -62,6 +70,9 @@ func init() {
 		fs.IntVar(&retryCount, "retry-count", 2, "retry count")
+		fs.BoolVar(&balancerEnabled, "balancer_enabled", false, "Whether to enable the tablet balancer to evenly spread query load")
+		fs.StringSliceVar(&balancerVtgateCells, "balancer_vtgate_cells", []string{}, "When in balanced mode, a comma-separated list of cells that contain vtgates (required)")
+		fs.StringSliceVar(&balancerKeyspaces, "balancer_keyspaces", []string{}, "When in balanced mode, a comma-separated list of keyspaces for which to use the balancer (optional)")
```
Yep exactly. That's how we did the rollout testing inside Slack -- started with a loadtest keyspace only and then moved to the one production keyspace where we saw the most imbalance, so we could focus on that without worrying about potential negative impact on the rest of the (thousands of) shards.

demmer commented:
@systay / @deepthi I finally came back from some vacation and put some time into this.

I fixed the rand issue, renamed the flags, and added the website change https://github.com/vitessio/vitess/pull/16351. 

LMK if there's anything more to do here. Thanks!

systay commented on go/vt/vtgate/balancer/balancer.go:
```diff
@@ -0,0 +1,367 @@
+/*
+Copyright 2023 The Vitess Authors.
```
nitpick: year is wrong

systay commented on go/vt/vtgate/balancer/balancer.go:
```diff
@@ -0,0 +1,367 @@
+	// for a given query to maintain the desired balanced allocation over multiple executions.
+	Pick(target *querypb.Target, tablets []*discovery.TabletHealth) *discovery.TabletHealth
+
+	// Balancer debug page request
```
Comment should have the following format 'DebugHandler ...' (with an optional leading article)

systay commented on go/vt/vtgate/balancer/balancer_test.go:
```diff
@@ -0,0 +1,409 @@
+// allow 2% fuzz
+const FUZZ = 2
+
+func fuzzyEquals(a, b int) bool {
```
nitpick: `stretchr/testify`, which is the test library we usually use, contains a method `assert.InDelta`, which you could use instead of this method

systay commented on go/vt/vtgate/balancer/balancer_test.go:
```diff
@@ -0,0 +1,409 @@
+			outflowPerCell := make(map[string]int)
+			for _, outflow := range a.Outflows {
+				for tabletCell, flow := range outflow {
+					if flow < 0 {
+						t.Errorf("balancer %v negative outflow", b.print())
+					}
```
```suggestion
					assert.GreaterOrEqual(t, 0, flow, b.print())
```

systay commented on go/vt/vtgate/balancer/balancer_test.go:
```diff
@@ -0,0 +1,409 @@
+		th := b.Pick(target, tablets)
+		allocation, totalAllocation := b.getAllocation(target, tablets)
+
+		if totalAllocation != ALLOCATION/2 {
+			t.Errorf("totalAllocation mismatch %s", b.print())
+		}
+
+		if allocation[th.Tablet.Alias.Uid] != ALLOCATION/4 {
+			t.Errorf("allocation mismatch %s, cell %s", b.print(), allTablets[0].Tablet.Alias.Cell)
+		}
+
+		if th.Tablet.Alias.Cell != "a" {
+			t.Errorf("shuffle promoted wrong tablet from cell %s", tablets[0].Tablet.Alias.Cell)
+		}
```
```suggestion
		assert.Equal(t, ALLOCATION/2, totalAllocation, "totalAllocation mismatch %s", b.print())
		assert.Equalf(t, ALLOCATION/4, allocation[th.Tablet.Alias.Uid], "allocation mismatch %s, cell %s", b.print(), allTablets[0].Tablet.Alias.Cell)
		assert.Equal(t, "a", th.Tablet.Alias.Cell, "shuffle promoted wrong tablet from cell %s", allTablets[0].Tablet.Alias.Cell)
```

systay commented on go/vt/vtgate/balancer/balancer_test.go:
```diff
@@ -0,0 +1,409 @@
+/*
+Copyright 2023 The Vitess Authors.
```
nitpic: year

systay reviewed (approved):
I had mostly nitpicks to add in this second round of review. The main algorithm looks good, is well commented and well tested.

systay reviewed (approved):
I had mostly nitpicks to add in this second round of review. The main algorithm looks good, is well commented and well tested.

venkatraju commented:
@systay @deepthi made a small PR to address the rest of the review comments. Could you please take a look?

deepthi commented:
@systay @rohit-nayak-ps do either of you think this needs a release note?

deepthi commented:
@venkatraju looks like this may need either a merge from main, or a simpler code change to pass linter checks.

venkatraju commented:
> @venkatraju looks like this may need either a merge from main, or a simpler code change to pass linter checks.

@deepthi Fixed the log line that was triggering the linter error.

deepthi commented:
> @systay @rohit-nayak-ps do either of you think this needs a release note?

I've added the label to block merging. This will be a nice thing to add to the release notes.
@venkatraju can you please add a section to the summary release notes doc for this feature? It should go into https://github.com/vitessio/vitess/blob/main/changelog/21.0/21.0.0/summary.md

venkatraju commented:
> @venkatraju can you please add a section to the summary release notes doc for this feature? It should go into https://github.com/vitessio/vitess/blob/main/changelog/21.0/21.0.0/summary.md

@deepthi added release notes. Could you review the text, and also comment on whether you want the new flags called out there?

(sorry about the messed up git history from a rebase that went wrong)

deepthi commented:
> > @venkatraju can you please add a section to the summary release notes doc for this feature? It should go into https://github.com/vitessio/vitess/blob/main/changelog/21.0/21.0.0/summary.md
> 
> @deepthi added release notes. Could you review the text, and also comment on whether you want the new flags called out there?
> 
> (sorry about the messed up git history from a rebase that went wrong)

This looks fine. Details on flags will be in the vtgate command reference (auto generated).
